WalletWasabi/Blockchain/TransactionBuilding/SmartCoinSelector.cs

```diff
@@ -17,18 +17,36 @@ public SmartCoinSelector(List<SmartCoin> unspentCoins)
 	}
 
 	private List<SmartCoin> UnspentCoins { get; }
+	private int IterationCount { get; set; }
 
-	/// <param name="unused">Unused parameter, make it an empty list.</param>
-	public IEnumerable<ICoin> Select(IEnumerable<ICoin> unused, IMoney target)
+	/// <param name="suggestion">We use this to detect if NBitcoin tries to suggest something different and indicate the error.</param>
+	/// <param name="target">Only <see cref="Money"/> type is really supported by this implementation.</param>
+	/// <remarks>Do not call this method repeatedly on a single <see cref="SmartCoinSelector"/> instance.</remarks>
+	public IEnumerable<ICoin> Select(IEnumerable<ICoin> suggestion, IMoney target)
 	{
-		var targetMoney = target as Money;
+		var targetMoney = (Money)target;
 
 		long available = UnspentCoins.Sum(x => x.Amount);
 		if (available < targetMoney)
 		{
 			throw new InsufficientBalanceException(targetMoney, available);
 		}
 
+		if (IterationCount > 500)
+		{
+			throw new TimeoutException("Coin selection timed out.");
+		}
+
+		// This is a hacky safety check to make sure the first few iterations can go through.
+		if (IterationCount > 10)
+		{
+			Money suggestedSum = Money.Satoshis(suggestion.Sum(c => (Money)c.Amount));
```

Comment:
Yeah, I don't get this either. If we'd do `if (IterationCount > 1)`, then it'd make sense. @molnard can explain what's going on here?

Reply:
I have changed it. I am not sure how the Select method is called from NBitcoin, so I wanted to make sure that the first 10 iterations will happen exactly the same way as before. Ofc if you are sure that it is repeating itself, it can be changed.

Reply:
I checked again NBitcoin code

https://github.com/MetacoSA/NBitcoin/blob/30772d6477ea8da08b41075ed1ae64db64ae4798/NBitcoin/TransactionBuilder.cs#L1801

I am not sure that assuming that: multiple iteration cannot happen under normar circumastences - in other words if Select is called multiple times we have to act and change the logic. 

Lucas also recommends to use the suggested coins after an iteration instead of the pre-set ones. 
https://github.com/zkSNACKs/WalletWasabi/pull/8367#discussion_r896806448

What if not only the coins are taken into account, there are a couple of logic connected with the fees in NBitcoin. Maybe we give back exactly the same solution but for the second time it will pass because NBitcoin is taking the fee differently. 

This could lead to changing a current behavior of the coinselector.

Reply:
It seems like NBitcoin is trying to be clever and iterate more on the transaction there is a case when it would find a good result is we would let it run. 

https://github.com/zkSNACKs/WalletWasabi/issues/8983#issuecomment-1324926573